# Workshop 8 - NumPy

In today's workshop, you will work with a reduced set of data from [[1]](#References). This data describes measurements of different dry beans, used for designing a system for classifying different types of beans. To run this workshop, you will need to download the file `beans.txt` from Blackboard and place it in the same folder as your workshop notebook.

Through the workshop, you will be expected to analyse the measured characteristics of the beans in the dataset:
- [Loading the dataset](#Loading-the-dataset)
- [Analysing the dataset](#Analysing-the-dataset)
  - [Array dimensions](#Array-dimensions)
    - [Exercise 1](#Exercise-1)
  - [Extreme values](#Extreme-values)
    - [Exercise 2](#Exercise-2)
    - [Exercise 3](#Exercise-3)
  - [Calculating new features](#Calculating-new-features)
    - [Exercise 4](#Exercise-5)
  - [Dataset statistics](#Dataset-statistics)
    - [Exercise 5](#Exercise-5)
    - [Exercise 6](#Exercise-6)
- [References](#References)

In [6]:
import numpy as np 

## Loading the dataset

The file `beans.txt` contains some measurements on 7 different types of beans. The below code snippet loads this data into a NumPy array, such that every row corresponds to a different bean.

The columns correspond to the following measurements of the bean: **area**, **perimiter**, **major axis length**, **minor axis length**. The last column represents the **class** (type of bean), and is encoded as an integer number.

The code for loading the dataset from the file `beans.txt` is provided. This code relies only on the material covered in the lectures and workshops so far. In the coming lectures, we will learn about libraries better suited for handling large data files.

In [7]:
with open('beans.txt', 'r') as beans_data:
    beans = np.array([line.strip().split(',') for line in beans_data], dtype='float')

To ensure we have correctly loaded the dataset, let's inspect the features (measurements) of the first bean:

In [8]:
print(beans[0])

[28395.           610.291        208.17811671   173.88874704
     0.        ]


## Analysing the dataset

### Array dimensions

Let's try and get a basic sense of the dataset by looking the number of samples it contains.

#### Exercise 1

How many different samples (beans) does this dataset contain?

In [11]:
print(beans.shape[0])

13611


### Extreme values

#### Exercise 2

Let's get a deeper look into the dataset. What is the area of the largest bean in the datset?

#### Exercise 3
 
Can you print all the features of the largest bean?

### Calculating new features

#### Exercise 4

Beans have an approximately ellipsoid shape. The ratio between the major and minor axis can tell us something about the shape of the bean. Calculate the **aspect ratio** between the major and minor axis of every bean and store it in a new array `aspect_ratio`.

**Extension:** Add this new informatio to the array representing the dataset as a new column. You will need [`numpy.transpose()`](https://numpy.org/doc/stable/reference/generated/numpy.transpose.html) and [`numpy.hstack()`](https://numpy.org/doc/stable/reference/generated/numpy.hstack.html#numpy.hstack).

### Dataset statistics

#### Exercise 5

First, calculate the mean and standard deviation of the aspect ratio of all the beans.

In [ ]:
mean_all = np.mean(beans[:, 5])
std_all = np.std(beans[:, 5])
print(mean_all, std_all)

#### Exercise 6

Now, let's compare the above global statistics for the dataset to those of class 3. Claculate the mean and standard deviation of the aspect ratio of all the beans of class 3 (you will need to use mask idexing).

You should be able to observe a difference between the mean and standard deviation of the measured characteristic between the whole dataset and class 3. **How would you interpret this difference** (especially the one in the standard deviation)?

### References

[1] KOKLU, M. and OZKAN, I.A., (2020), "Multiclass Classification of Dry Beans Using Computer Vision and Machine Learning Techniques." Computers and Electronics in Agriculture, 174, 105507. DOI: https://doi.org/10.1016/j.compag.2020.105507